### PR TITLE
Adding podaffinity and resourcerequirments to IG Spec

### LINKED ISSUE #2455: Add support for resources/limits and affinity to IG spec.
/kind feature

**Describe the solution you'd like**
IG’s router resources (MEM and CPU) are only configurable at global level using `router` section of `inference-service` config map.
Make router’s resources  configurable per IG like we can set per ISVC.

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

### LINKED ISSUE #2696: Allow setting affinity spec for inference graph
/kind feature

**Describe the solution you'd like**
Today inference services as part of its pod spec, allows to introduce affinity, which allows for HA across availability zones. As minReplicas will be introduced as part of #2456, affinity becomes more important, so that all routers dont end up on the same node. 

Also, since InferenceService can be distributed across nodes (use pod-anti affinity and AZ topology key) with a high chance of all Inference services part of the graph scheduled on the same node, it make sense for the routers to also be distributed.

The below is what i am envisioning using minReplicas as well as pod anti affinity

ig-affinity

**Anything else you would like to add:**
Min replicas for IG router is being solved by: #2456

### PR DESCRIPTION
**Pull request by rachitchauhan43**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #2455 
Fixes #2696 

**Type of changes**
Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**Feature/Issue validation/testing**:

Please describe the tests that you ran to verify your changes and relevent result summary. Provide instructions so it can be reproduced.
Please also list any relevant details for your test configuration.

- [ ] Test A
- [ ] Test B

- Logs

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Checklist**:

- [ ] Have you added unit/e2e tests that prove your fix is effective or that this feature works?
- [ ] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

**Release note**:

```release-note

```

### CODE REVIEW

File: pkg/apis/serving/v1alpha1/inference_graph.go [hunk 1]
```diff
@@ -45,6 +46,10 @@ type InferenceGraphSpec struct {
 	// Map of InferenceGraph router nodes
 	// Each node defines the router which can be different routing types
 	Nodes map[string]InferenceRouter `json:"nodes"`
+	// +optional
+	Resources corev1.ResourceRequirements `json:"resources,omitempty" protobuf:"bytes,8,opt,name=resources"`
```
Comment (yuzisun) [inline pkg/apis/serving/v1alpha1/inference_graph.go]: Not sure if we need the protobuf def as we did not define for the other graph fields, also 8 here means the ordering of the fields on container spec but we do not include any of the other fields.

**Comment by rachitchauhan43** [comment]
@yuzisun : As discussed, added tests for IG controller. We can add more as we add more features.

File: pkg/controller/v1alpha1/inferencegraph/controller_test.go [hunk 2]
```diff
@@ -0,0 +1,461 @@
+package inferencegraph
```
Comment (yuzisun) [inline pkg/controller/v1alpha1/inferencegraph/controller_test.go:17]: Missing the Apache license

File: pkg/controller/v1alpha1/inferencegraph/controller_test.go [hunk 3]
```diff
@@ -0,0 +1,461 @@
...
+					        - weight: 100
+					          podAffinityTerm:
+					            labelSelector:
+					              matchExpressions:
+					                - key: serving.kserve.io/inferencegraph
+					                  operator: In
+					                  values:
+					                    - usual-flow
+					            topologyKey: topology.kubernetes.io/zone
+				*/
```
Comment (yuzisun) [inline pkg/controller/v1alpha1/inferencegraph/controller_test.go]: Can be removed ?

File: pkg/controller/v1alpha1/inferencegraph/suite_test.go [hunk 4]
```diff
@@ -0,0 +1,105 @@
+package inferencegraph
```
Comment (yuzisun) [inline pkg/controller/v1alpha1/inferencegraph/suite_test.go:17]: missing Apache license

File: pkg/controller/v1alpha1/inferencegraph/suite_test.go [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+
+import (
+	"context"
+	kfservingv1alpha1 "github.com/kserve/kserve/pkg/apis/serving/v1alpha1"
+	"github.com/kserve/kserve/pkg/apis/serving/v1beta1"
+	"github.com/kserve/kserve/pkg/constants"
+	//"github.com/kserve/kserve/pkg/controller/v1alpha1/inferencegraph"
```
Comment (yuzisun) [inline pkg/controller/v1alpha1/inferencegraph/suite_test.go]: delete this line?

File: pkg/controller/v1alpha1/inferencegraph/controller_test.go (hunk 2 above)
Comment (rachitchauhan43) [reply pkg/controller/v1alpha1/inferencegraph/controller_test.go:17]: Added

File: pkg/controller/v1alpha1/inferencegraph/controller_test.go (hunk 3 above)
Comment (rachitchauhan43) [reply pkg/controller/v1alpha1/inferencegraph/controller_test.go]: Done

File: pkg/controller/v1alpha1/inferencegraph/suite_test.go (hunk 4 above)
Comment (rachitchauhan43) [reply pkg/controller/v1alpha1/inferencegraph/suite_test.go:17]: Added

File: pkg/controller/v1alpha1/inferencegraph/suite_test.go (hunk 5 above)
Comment (rachitchauhan43) [reply pkg/controller/v1alpha1/inferencegraph/suite_test.go]: Done

File: pkg/apis/serving/v1alpha1/inference_graph.go (hunk 1 above)
Comment (rachitchauhan43) [reply pkg/apis/serving/v1alpha1/inference_graph.go]: Addressed.

**Comment by yuzisun** [comment]
@rachitchauhan43 Awesome work! 

/lgtm
/approve

**Pull request merged**